### ISSUE TITLE
Use keras tokenizer

### ISSUE THREAD

**Issue by pnrobinson** [monarch-initiative/N2V#45]
Replace the self-made tokenizer.
This will reduce the verbosity of our code.
See example 4 here:
https://www.programcreek.com/python/example/106871/keras.preprocessing.text.Tokenizer

**Comment by deepakunni3** [comment]
@pnrobinson Should I start with this? Or are you working on this as part of https://github.com/monarch-initiative/N2V/pull/47?

**Comment by pnrobinson** [comment]
@deepakunni3 -- please go ahead with this! I think it is localized to the text_encoder file, so there should be no problem.

**Comment by deepakunni3** [comment]
Awesome! I'll get started on it.

**Comment by callahantiff** [comment]
@deepakunni3 - in case it's helpful, I also found [this](https://keras.io/preprocessing/text/) to be super helpful. Looks like its default output lines up with what we need to.

**Comment by deepakunni3** [comment]
@callahantiff Thank you for the link. Yes, this is definitely helpful!

**Comment by deepakunni3** [comment]
I have been going through the Keras Preprocessing documentation and I wanted to make a few observations:
- Keras does not handle filtering of stop words. So we have to do it manually (or use nltk)
- Keras does not undo contractions
- In the word dictionary, Keras reserves the 0th index for padding and uses the 1st index for `UNK` token

Now, keeping these limitations in mind I have a PR (https://github.com/monarch-initiative/N2V/pull/51) that uses `keras_preprocessing.text.Tokenizer` in TextEncoder.build_dataset() method.

**Comment by callahantiff** [comment]
I'm also happy to help debug, let me know if I can help!

**Comment by callahantiff** [comment]
I have been playing around with a python library called [`ktext`](https://github.com/hamelsmu/ktext). it's specifically designed to preprocess text for input into keras/tensorflow and it can be parallelized.

I tested it out with some minor tweaks (mainly for situations where don't want padding) and it seems to run fine and does all of the preprocessing we want plus a lot more (it includes a ton of awesome funcs for cleaning). If we adopted this, I think we could essentially replace the `text_encoder.py` script with a small method.

Here is the [tutorial](https://github.com/hamelsmu/ktext/blob/master/notebooks/Tutorial.ipynb) provided in the repo if you are curious.

Happy to provide more details on how I ran it, if there is interest. Here is a small snippet if you are curious.

```python
# import ktext
from ktext.preprocess import processor

# read in data
data_sentences = open('Emails.csv').readlines()

# run tokenizer
issue_body_proc = processor(keep_n=50000)
train_result = issue_body_proc.fit_transform(data_sentences)

# process output data
data1_sents_no_padding = [list(filter(None, list(sent))) for sent in train_result if len(set(list(sent))) > 1]

# convert numpy array to nested list
data1_doc_no_pad = [x for y in data1_sents_no_padding for x in y]

# get output needed for w2v
count1 = list(issue_body_proc.token_count_pandas().itertuples(index=True, name=None))
dictionary1 = issue_body_proc.token2id
reverse_dictionary1 = issue_body_proc.id2token

# run cbow
model_cb = ContinuousBagOfWordsWord2Vec(data1_doc_no_pad, worddictionary=dictionary1,
                                        reverse_worddictionary=reverse_dictionary1)
model_cb.add_display_words(count1)
model_cb.train(display_step=1000)
```

**Comment by deepakunni3** [comment]
@callahantiff This is brilliant! 
This was the original intent of using Keras Tokenizer - to reduce verbosity. But looks like ktext is a better alternative here.

**Comment by pnrobinson** [comment]
I think that the text_encoder class does this, we are all set!
